enhancements/alerting/20230124-routing-metrics-backend.md

```diff
@@ -0,0 +1,144 @@
+// powered by Cortex AlertManager
+service AlertingAdminServer {
+  //...
+  rpc SyncConfig(UserConfigRequest) returns (google.protobuf.Empty){}
+
+  rpc ListSyncConfigs(ListSyncedConfigRequest) returns (syncConfigList){}
+
+  rpc ImportFile(UserFileImportRequest) returns (google.protobuf.Empty) {}
+
+  rpc ListImportedFiles(ListImportedFiles) returns (google.protobuf.Empty) {}
+  //....
+}
+
+message SyncConfig{
+  string clusterKey = 1;
+  byte config = 2;
+}
+
+message SyncConfigList{
+  map<string,SyncConfig> items = 1;
+}
+
+message UserFileImportRequest {
+  string clusterKey = 1;
+  byte data = 2;
+  destinationPath string = 3;
+}
+
+message ListSyncedConfigRequest{
+  // cluster keys
+  repeated string items = 1;
+}
+
+message ListImportedFiles{
+  // cluster keys
+  repeated string items = 1;
+}
+```
+
+intended to be called from the CLI at least initially:
+
+- `opni alerting sync <filename> --cluster <cluster-name or id>`
+
+many production settings require setting explicit files:
+
+- `opni alerting sync import <filename> --cluster <cluster-name> --destination-path`
+
+<hr/>
+
+- the current `AlertEndpoint` tag system (list of strings) will be replaced with prometheus label matchers
+
+Some additional notes:
+
+- Syncing user configurations will return a validation error if the configuration requires files that are not found -- see above
+
+<hr/d>
+
+- Endpoints found in the sync process will be listed in the Endpoints UI, but will appear as **Read only** && **Synced** via labels
+- Attempting to edit the endpoints through the admin UI will result in a `validation error` prompting the user to edit the configs that have already synced directly instead.
+
+<hr/>
+
+- Conditions that have labels matching the endpoints will have these endpoints attached to them automatically, meaning Opni internal conditions will be able to use the synced endpoints
+
+### UI
+
+- Endpoint pages for each endpoint type
+- Tags migrated to label matcher system with operations `==`,`!=`,`~=`,`~!=` for both endpoints & conditions
+
+## Acceptance criteria:
+
+- [ ] Reach endpoint parity with AlertManager by implementing Opni receiver functionality on remaining AlertManager receiver(s)
+  - [ ] `Webhook`
+  - [ ] `Amazon SNS`
+  - [ ] `OpsGenie`
+  - [ ] `Pushover`
+  - [ ] `Victorops`
+  - [ ] `Wechat`
+  - [ ] `Telegram`
+  - [ ] `Webex`
+- [ ] Sync user configurations to Opni Router & Attach Cortex to them
+- [ ] Import filesystem files for user configurations to Alerting Cluster
+- [ ] Extend Opni Endpoint protocol buffer messages to include Prometheus `label matchers`
+  - [] `==`, `!=`
+  - [] `~=`, `~!=`
+- [ ] Optimize `label matchers` when building Opni Router
+
+## Supporting documents:
+
+## Dependencies:
+
+- https://github.com/rancher/opni/pull/942
+
+## Risks and contingencies:
+
+| Risk | Contingency |
+| Some AlertManager endpoints may require premium licenses for functionality | Request license through SUSE |
```

Comment:
Check to see if alertmanager/cortex/etc tests these already, if so we can skip

Reply:
The issue is that adding a config that violates structural or runtime assumptions may break all the runtime alertmanager functionality - with very little indication of what is wrong.

My philosophy is we shouldn't let the users hang themselves, if possible.

One other option is to buffer the error and debug logs from within the AlertManager process and group these logs by endpoint, then expose these via an API.  But that will be stateful and still doesn't help when the process exits abnormally.

The core requirement of Opni-Alerting is not presupposing any knowledge about AlertManager or Opni backends, I'm still leaning towards manually testing them and having as excellent as possible input validation to prevent these issues. Thoughts?

Reply:
That sounds reasonable to me. :+1: